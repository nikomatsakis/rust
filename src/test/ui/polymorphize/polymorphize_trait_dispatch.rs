// Tests focused on trait dispatch.
//
// compile-flags: -Zpolymorphize=0 -Zpolymorphize-dump -Zpolymorphize-duplicates
// compile-pass

#![allow(warnings)]

fn depend_trait_dispatch<T: Clone>(t: &T) -> T {
    //~^ ERROR some polymorphic dependencies found
    t.clone()
}

fn depend_trait_dispatch_indirect<T: Clone>(t: &T) -> T {
    //~^ ERROR some polymorphic dependencies found
    depend_trait_dispatch(t)
}

fn depend_trait_dispatch_ref_indirect<T: Clone>(t: &T) -> &T {
    //~^ ERROR some polymorphic dependencies found
    depend_trait_dispatch::<&T>(&t)
}

fn main() {
    //~^ ERROR no polymorphic dependencies found

    // Invoke each function so that they are considered by the collector.
    depend_trait_dispatch::<u32>(&3);
    depend_trait_dispatch::<u16>(&3);

    depend_trait_dispatch_indirect::<u32>(&3);
    depend_trait_dispatch_indirect::<u16>(&3);

    depend_trait_dispatch_ref_indirect::<u32>(&3);
    depend_trait_dispatch_ref_indirect::<u16>(&3);
}
