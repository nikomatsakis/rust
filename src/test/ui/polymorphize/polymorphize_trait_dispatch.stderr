warning: some polymorphic dependencies found
  --> $DIR/polymorphize_trait_dispatch.rs:8:1
   |
LL | / fn depend_trait_dispatch<T: Clone>(t: &T) -> T {
LL | |     //~^ ERROR some polymorphic dependencies found
LL | |     t.clone()
   | |     --------- depends on `TraitMethod(DefId(2/0:1278 ~ core[8468]::clone[0]::Clone[0]::clone[0]), [T])`
LL | | }
   | |_^

warning: some polymorphic dependencies found
  --> $DIR/polymorphize_trait_dispatch.rs:13:1
   |
LL |       t.clone()
   |       --------- depends on `TraitMethod(DefId(2/0:1278 ~ core[8468]::clone[0]::Clone[0]::clone[0]), [T])`
...
LL | / fn depend_trait_dispatch_indirect<T: Clone>(t: &T) -> T {
LL | |     //~^ ERROR some polymorphic dependencies found
LL | |     depend_trait_dispatch(t)
   | |     ------------------------ invokes `DefId(0/0:3 ~ polymorphize_trait_dispatch[317d]::depend_trait_dispatch[0])` with substitutions `[T]`
LL | | }
   | |_^

warning: some polymorphic dependencies found
  --> $DIR/polymorphize_trait_dispatch.rs:18:1
   |
LL |       t.clone()
   |       ---------
   |       |
   |       depends on `OtherMethod([ReErased, T])`
   |       invokes `DefId(2/0:1292 ~ core[8468]::clone[0]::impls[0]::{{impl}}[3]::clone[0])` with substitutions `[ReErased, T]`
...
LL | / fn depend_trait_dispatch_ref_indirect<T: Clone>(t: &T) -> &T {
LL | |     //~^ ERROR some polymorphic dependencies found
LL | |     depend_trait_dispatch::<&T>(&t)
   | |     ------------------------------- invokes `DefId(0/0:3 ~ polymorphize_trait_dispatch[317d]::depend_trait_dispatch[0])` with substitutions `[&T]`
LL | | }
   | |_^

warning: no polymorphic dependencies found
  --> $DIR/polymorphize_trait_dispatch.rs:23:1
   |
LL | / fn main() {
LL | |     //~^ ERROR no polymorphic dependencies found
LL | | }
   | |_^

note: Monomorphized items: 9

note: Monomorphized size : 32

note: New total items    : 9 (100%)

note: New estimated size : 32 (100%)

